esp_commands/src/esp_commands.c

```diff
@@ -0,0 +1,720 @@
+        cmd_set->cmd_ptr_set = heap_caps_malloc(alloc_cmd_ptrs_size, s_config.heap_caps_used);
+        if (!cmd_set->cmd_ptr_set) {
+            return ESP_ERR_NO_MEM;
+        } else {
+            /* copy the temp set of pointer in to the final destination */
+            memcpy(cmd_set->cmd_ptr_set, cmd_ptrs, alloc_cmd_ptrs_size);
+            cmd_set->cmd_set_size = cmd_count;
+        }
+    }
+    return ESP_OK;
+}
+
+esp_command_set_handle_t esp_commands_create_cmd_set(const char **cmd_set, const size_t cmd_set_size, esp_commands_get_field_t get_field)
+{
+    if (!cmd_set || cmd_set_size == 0) {
+        return NULL;
+    }
+
+    esp_command_sets_t *cmd_ptr_sets = heap_caps_malloc(sizeof(esp_command_sets_t), s_config.heap_caps_used);
+    if (!cmd_ptr_sets) {
+        return NULL;
+    }
+
+
+    esp_command_t *static_cmd_ptrs_temp[ESP_COMMANDS_COUNT];
+    esp_command_t *dynamic_cmd_ptrs_temp[esp_dynamic_commands_get_number_of_cmd()];
+    create_cmd_set_ctx_t ctx = {
+        .cmd_set_name = NULL,
+        .get_field = get_field,
+        .static_cmd_ptrs = static_cmd_ptrs_temp,
+        .static_cmd_count = 0,
+        .dynamic_cmd_ptrs = dynamic_cmd_ptrs_temp,
+        .dynamic_cmd_count = 0
+    };
+
+    /* populate the temporary cmd pointer sets */
+    for (size_t i = 0; i < cmd_set_size; i++) {
+        ctx.cmd_set_name = cmd_set[i];
+        go_through_commands(NULL, &ctx, fill_temp_set_info);
+    }
+
+    /* if no static command was found, return a static set with 0 items in it */
+    esp_err_t ret_val = update_cmd_set_with_temp_info(&cmd_ptr_sets->static_set,
+                        ctx.static_cmd_count,
+                        ctx.static_cmd_ptrs);
+    if (ret_val == ESP_ERR_NO_MEM) {
+        free(cmd_ptr_sets);
+        return NULL;
+    }
+
+    /* if no dynamic command was found, return a dynamic set with 0 items in it */
+    ret_val = update_cmd_set_with_temp_info(&cmd_ptr_sets->dynamic_set,
+                                            ctx.dynamic_cmd_count,
+                                            ctx.dynamic_cmd_ptrs);
+    if (ret_val == ESP_ERR_NO_MEM) {
+        free(cmd_ptr_sets->static_set.cmd_ptr_set);
+        free(cmd_ptr_sets);
+        return NULL;
+    }
+
+    return (esp_command_set_handle_t)cmd_ptr_sets;
+}
+
+esp_command_set_handle_t esp_commands_concat_cmd_set(esp_command_set_handle_t cmd_set_a, esp_command_set_handle_t cmd_set_b)
+{
+    if (!cmd_set_a && !cmd_set_b) {
+        return NULL;
+    } else if (cmd_set_a && !cmd_set_b) {
+        return cmd_set_a;
+    } else if (!cmd_set_a && cmd_set_b) {
+        return cmd_set_b;
+    }
+
+    /* Reaching this point, both cmd_set_a and cmd_set_b are set.
+     * Create a new cmd_set that can host the items from both sets,
+     * assign the items to the new set and free the input sets */
+    esp_command_sets_t *concat_cmd_sets = heap_caps_malloc(sizeof(esp_command_sets_t), s_config.heap_caps_used);
```

Comment:
We defined our own malloc but still use this way of doing it a few places. Maybe ocnsider if the custom malloc function is necessary